icicle-compiler/src/Icicle/Sea/Header.hs

```diff
@@ -0,0 +1,557 @@
+ppOutputId =
+  ppRendered renderOutputId
+
+pMeltedOutput :: Aeson.Value -> Aeson.Parser (OutputId, MeltedType)
+pMeltedOutput =
+  Aeson.withObject "(OutputId, MeltedType)" $ \o ->
+    (,)
+      <$> withKey "output_id" o pOutputId
+      <*> withKey "type" o pMeltedType
+
+ppMeltedOutput :: (OutputId, MeltedType) -> Aeson.Value
+ppMeltedOutput (oid, mtype) =
+  Aeson.object [
+      "output_id" .=
+        ppOutputId oid
+    , "type" .=
+        ppMeltedType mtype
+    ]
+
+pNamedType :: Aeson.Value -> Aeson.Parser (SeaName, ValType)
+pNamedType =
+  Aeson.withObject "(SeaName, ValType)" $ \o ->
+    (,)
+      <$> withKey "name" o pSeaName
+      <*> withKey "type" o pValType
+
+ppNamedType :: (SeaName, ValType) -> Aeson.Value
+ppNamedType (name, vtype) =
+  Aeson.object [
+      "name" .=
+        ppSeaName name
+    , "type" .=
+        ppValType vtype
+    ]
+
+pKernel :: Aeson.Value -> Aeson.Parser Kernel
+pKernel =
+  Aeson.withObject "Kernel" $ \o ->
+    Kernel
+      <$> withKey "kernel_id" o pKernelId
+      <*> withKey "resumables" o (pList "Array (SeaName, ValType)" pNamedType)
+      <*> withKey "outputs" o (pList "Array (OutputId, MeltedType)" pMeltedOutput)
+
+ppKernel :: Kernel -> Aeson.Value
+ppKernel x =
+  Aeson.object [
+      "kernel_id" .=
+        ppKernelId (kernelId x)
+    , "resumables" .=
+        ppList ppNamedType (kernelResumables x)
+    , "outputs" .=
+        ppList ppMeltedOutput (kernelOutputs x)
+    ]
+
+pCluster :: Aeson.Value -> Aeson.Parser Cluster
+pCluster =
+  Aeson.withObject "Cluster" $ \o ->
+    Cluster
+      <$> withKey "cluster_id" o pClusterId
+      <*> withKey "input_id" o pInputId
+      <*> withKey "input_type" o pValType
+      <*> withKey "input_vars" o (pList "Array (SeaName, ValType)" pNamedType)
+      <*> withKey "time_var" o pSeaName
+      <*> withKey "kernels" o (pNonEmpty "NonEmpty Kernel" pKernel)
+
+ppCluster :: Cluster -> Aeson.Value
+ppCluster x =
+  Aeson.object [
+      "cluster_id" .=
+        ppClusterId (clusterId x)
+    , "input_id" .=
+        ppInputId (clusterInputId x)
+    , "input_type" .=
+        ppValType (clusterInputType x)
+    , "input_vars" .=
+        ppList ppNamedType (clusterInputVars x)
+    , "time_var" .=
+        ppSeaName (clusterTimeVar x)
+    , "kernels" .=
+        ppNonEmpty ppKernel (clusterKernels x)
+    ]
+
+pHeader :: Aeson.Value -> Aeson.Parser Header
+pHeader =
+  Aeson.withObject "Header" $ \o -> do
+    expectKey "header" "version" o pText "v0"
+    Header
+      <$> withKey "fingerprint" o pFingerprint
+      <*> withKey "clusters" o (pList "Map InputId Cluster" pCluster)
+
+ppHeader :: Header -> Aeson.Value
+ppHeader x =
+  Aeson.object [
+      "version" .=
+        ppText "v0"
+    , "fingerprint" .=
+        ppFingerprint (headerFingerprint x)
+    , "clusters" .=
+        ppList ppCluster (headerClusters x)
+    ]
+
+headerKeyOrder :: [Text]
+headerKeyOrder = [
+  -- Header
+    "version"
+  , "fingerprint"
+  , "clusters"
```

Comment:
Just a string, I was thinking of putting the icicle version in there, but if we call this from blizzard it would be the blizzard version instead.

We can only get the version/git hash in the main executable.